projects/addon-preview/components/preview/preview.component.ts

```diff
@@ -0,0 +1,253 @@
+        if (contentResizeEntries.length === 0) {
+            return;
+        }
+
+        const {width, height} = contentResizeEntries[0].contentRect;
+
+        this.refresh(width, height);
+    }
+
+    reset() {
+        this.zoom$.next(this.minZoom);
+        this.coordinates$.next(EMPTY_COORDINATES);
+    }
+
+    private get offsets(): {offsetX: number; offsetY: number} {
+        const offsetX = ((this.zoom$.value - this.minZoom) * this.width) / 2;
+        const offsetY = ((this.zoom$.value - this.minZoom) * this.height) / 2;
+
+        return {offsetX, offsetY};
+    }
+
+    private calculateMinZoom(
+        contentHeight: number,
+        contentWidth: number,
+        boxHeight: number,
+        boxWidth: number,
+    ): number {
+        const bigSize =
+            contentHeight > boxHeight * INITIAL_SCALE_COEF ||
+            contentWidth > boxWidth * INITIAL_SCALE_COEF;
+
+        return bigSize
+            ? round(
+                  Math.min(
+                      (this.elementRef.nativeElement.clientHeight * INITIAL_SCALE_COEF) /
+                          contentHeight,
+                      (this.elementRef.nativeElement.clientWidth * INITIAL_SCALE_COEF) /
+                          contentWidth,
+                  ),
+                  2,
+              )
+            : 1;
+    }
+
+    private initClickSubscription(element: HTMLElement) {
+        dragAndDropFrom(element)
+            .pipe(
+                filter(() => this.zoomable),
+                pairwise(),
+                filter(
+                    ([previous, current]) =>
+                        previous !== null &&
+                        current !== null &&
+                        previous.stage === TuiDragStage.Start &&
+                        current.stage === TuiDragStage.End,
+                ),
+                takeUntil(this.destroy$),
+            )
+            .subscribe(([{event}]) => {
+                this.processZoom(
+                    event.clientX,
+                    event.clientY,
+                    this.zoom$.value > this.minZoom ? -0.5 : 0.5,
+                );
+            });
+    }
+
+    private refresh(width: number, height: number) {
+        this.width = width;
+        this.height = height;
+        this.minZoom = this.calculateMinZoom(
+            height,
+            width,
+            this.elementRef.nativeElement.clientHeight,
+            this.elementRef.nativeElement.clientWidth,
+        );
+        this.zoom$.next(this.minZoom);
+        this.rotation$.next(0);
+    }
+
+    private processZoom(clientX: number, clientY: number, delta: number) {
+        const oldScale = this.zoom$.value;
+        const newScale = clamp(this.zoom$.value + delta, this.minZoom, 2);
+
+        const center = this.getScaleCenter(
+            {clientX, clientY},
+            this.coordinates$.value,
+            this.zoom$.value,
+        );
+
+        this.zoom$.next(newScale);
```

Comment:
Let's move it down to the other `next` so the resulting action is all in the same place

Reply:
We have to update zoom first and then compute new coordinates, so the order matters. ( Maybe split this method in two?